V malob/docs (#8)
akv-platform/setup-java: docs/usage.md

- [Installing custom Java architecture](#Installing-custom-Java-architecture)
- [Installing custom Java distribution from local file](#Installing-Java-from-local-file)
- [Testing against different Java versions](#Testing-against-different-Java-versions)
- [Testing against different Java distributions](#Testing-against-different-Java-distributions)
- [Publishing using Apache Maven](#Publishing-using-Apache-Maven)
- [Publishing using Gradle](#Publishing-using-Gradle)

See [action.yml](../action.yml) for more details on task inputs.

## Selecting Java distribution
Input `distribution` is mandatory and should be provided to use action. See [Supported distributions](../README.md#Supported-distributions) for the list of available options.

### Adoptium
```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: 'adoptium'
    java-version: '11.x'
- run: java -cp java HelloWorldApp
```

### Zulu
```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: 'zulu'
    java-version: '11.x'
    java-package: jdk # optional (jdk, jre or jdk+fx) - defaults to jdk
- run: java -cp java HelloWorldApp
```

> To use 

Reviewer: please, remove

  - [Supported version syntax](#Supported-version-syntax)
  - [Zulu](#Zulu)
  - [Adoptium](#Adoptium)
- [Installing custom Java package type](#Installing-custom-Java-package-type)
- [Installing custom Java architecture](#Installing-custom-Java-architecture)
- [Installing custom Java distribution from local file](#Installing-Java-from-local-file)
- [Testing against different Java versions](#Testing-against-different-Java-versions)
- [Testing against different Java distributions](#Testing-against-different-Java-distributions)
- [Publishing using Apache Maven](#Publishing-using-Apache-Maven)
- [Publishing using Gradle](#Publishing-using-Gradle)

See [action.yml](../action.yml) for more details on task inputs.

## Selecting Java distribution
Input `distribution` is mandatory and should be provided to use action. See [Supported distributions](../README.md#Supported-distributions) for the list of available options.

```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: '<distribution>' # See 'Supported distributions' for available distribution
    java-version: '11.x'
- run: java -cp java HelloWorldApp

> ```

Reviewer: this section almost fully duplicates Basic section from the main README, but we announced it as Advanced usage. 
I'd suggest removing it from here and moving Adoptium and Zulu examples here instead (even above the "Supported version syntax" section) because the main header is "Selecting Java distribution". 
And section "Supported version syntax" I'd make a 2-level header, because it doesn't really relate to distributors.

TO-DO: Clarify docs about using syntax with 4 digits  

### Zulu
```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: 'zulu'
    java-version: '11.x'
    java-package: jdk # optional (jdk or jre) - defaults to jdk
- run: java -cp java HelloWorldApp
```

### Adoptium
```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: 'adoptium'
    java-version: '11.x'
- run: java -cp java HelloWorldApp
```

## Installing custom Java package type
```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: '<distribution>'
    java-version: '11.x'
    java-package: jdk # optional (jdk or jre) - defaults to jdk
- run: java -cp java HelloWorldApp
```

## Installing custom Java architecture

```yaml
steps:
- uses: actions/checkout@v2
- uses: actions/setup-java@v2-preview
  with:
    distribution: '<distribution>'
    java-version: '11.x'
    architecture: x86 # optional - defaults to x64
- run: java -cp java HelloWorldApp
```

## Installing Java from local file

> If your use-case requires using custom distribution or version that is not provided in action by default, you can download it manually and then action will take care about installation and caching version on VM:

Reviewer: ```suggestion
If your use-case requires custom distribution or version that is not provided by action, you can download it manually and then action will take care about installation and caching of this version on VM:
```
not sure what should it mean "by default", distribution either provided or not, and we have no default distribution.

TO-DO: Fix example
```yaml
steps:
- run: |
    download_url="https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/download/jdk-11.0.10%2B9/OpenJDK11U-jdk_x64_mac_hotspot_11.0.10_9.pkg"
    wget $download_url
- uses: actions/setup-java@v2-preview
  with:
    distribution: 'jdkFile'

>     jdkFile: runner.temp # Optional - jdkFile to install java from. Useful for distributions not provided by default

Reviewer: ```suggestion
    jdkFile: runner.temp # Optional - jdkFile to install java from.
```
removing it because we already said above about distributions not provided by action

    java-version: '11.0.0'
    architecture: x64
    
- run: java -cp java HelloWorldApp
```

## Testing against different Java versions
```yaml
jobs:
  build:
    runs-on: ubuntu-20.04
    strategy:
      matrix:
        java: [ '8', '11', '13', '15' ]
    name: Java ${{ matrix.Java }} sample
    steps:
      - uses: actions/checkout@v2
      - name: Setup java
        uses: actions/setup-java@v2-preview
        with:
          distribution: '<distribution>'
          java-version: ${{ matrix.java }}
      - run: java -cp java HelloWorldApp
```

## Testing against different Java distributions

> **NOTE:** Please pay attention that different distributors provide different list of available versions / supported configurations.

Reviewer: ```suggestion
**NOTE:** The different distributors can provide discrepant list of available versions / supported configurations. Please refer to the official documentation to see the list of supported versions.
```
I like more wording from the main README

        java: [ '8', '11', '13', '15' ]
    name: Java ${{ matrix.Java }} (${{ matrix.distribution }}) sample
    steps:
      - uses: actions/checkout@v2
      - name: Setup java
        uses: actions/setup-java@v2-preview
        with:
          distribution: ${{ matrix.distribution }}
          java-version: ${{ matrix.java }}
      - run: java -cp java HelloWorldApp
```

## Publishing using Apache Maven

### Extra setup for pom.xml:

The Maven GPG Plugin configuration in the pom.xml file should contain the following structure to avoid possible issues like `Inappropriate ioctl for device` or `gpg: signing failed: No such file or directory`:

```xml
<configuration>
  <!-- Prevent gpg from using pinentry programs -->
  <gpgArguments>
    <arg>--pinentry-mode</arg>
    <arg>loopback</arg>
  </gpgArguments>
</configuration>
```
GPG 2.1 requires `--pinentry-mode` to be set to `loopback` in order to pick up the `gpg.passphrase` value defined in Maven `settings.xml`.

### Yaml example:
```yaml
jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up JDK 8.x
      uses: actions/setup-java@v2-preview
      with:

>         distribution: 'zulu'

Reviewer: in all examples we use `'<distribution>'` but `'zulu'` here. Probably makes sense update it to `'<distribution>'` too

### GPG

If `gpg-private-key` input is provided, the private key will be written to a file in the runner's temp directory, the private key file will be imported into the GPG keychain, and then the file will be promptly removed before proceeding with the rest of the setup process. A cleanup step will remove the imported private key from the GPG keychain after the job completes regardless of the job status. This ensures that the private key is no longer accessible on self-hosted runners and cannot "leak" between jobs (hosted runners are always clean instances).

**GPG key should be exported by: `gpg --armor --export-secret-keys YOUR_ID`**

See the help docs on [Publishing a Package](https://help.github.com/en/github/managing-packages-with-github-packages/configuring-apache-maven-for-use-with-github-packages#publishing-a-package) for more information on the `pom.xml` file.

## Publishing using Gradle
```yaml
jobs:

  build:
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2

    - name: Set up JDK 1.8

>       uses: actions/setup-java@v1

Reviewer: this should be updated to proper usage

        PASSWORD: ${{ secrets.GITHUB_TOKEN }}
```

***NOTE: The `USERNAME` and `PASSWORD` need to correspond to the credentials environment variables used in the publishing section of your `build.gradle`.***

See the help docs on [Publishing a Package with Gradle](https://help.github.com/en/github/managing-packages-with-github-packages/configuring-gradle-for-use-with-github-packages#example-using-gradle-groovy-for-a-single-package-in-a-repository) for more information on the `build.gradle` configuration file.

## Apache Maven with a settings path

When using an Actions self-hosted runner with multiple shared runners the default `$HOME` directory can be shared by a number runners at the same time which could overwrite existing settings file. Setting the `settings-path` variable allows you to choose a unique location for your settings file.

```yaml
jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up JDK 1.8 for Shared Runner
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
        server-id: github # Value of the distributionManagement/repository/id field of the pom.xml

>         settings-path: ${{ github.workspace }} # location for the settings.xml file

Reviewer: let's get rid of 1.8 notation and add distribution input
